test/broadway_cloud_pub_sub/google_api_client_test.exs

```diff
@@ -336,20 +338,69 @@ defmodule BroadwayCloudPubSub.GoogleApiClientTest do
       assert url == "https://pubsub.googleapis.com/v1/projects/foo/subscriptions/bar:acknowledge"
     end
 
-    test "retries", %{opts: base_opts} do
+    test "retries with error response", %{opts: opts} do
       test_pid = self()
+      max_retries = opts[:retry][:max_retries]
       {:ok, counter} = Agent.start_link(fn -> 0 end)
 
       Tesla.Mock.mock(fn %{method: :post} ->
-        if Agent.get_and_update(counter, &{&1, &1 + 1}) < 3 do
+        if Agent.get_and_update(counter, &{&1, &1 + 1}) < max_retries do
           send(test_pid, :pong)
           {:error, %Tesla.Env{status: 503}}
         else
           {:ok, %Tesla.Env{status: 200, body: "{}"}}
         end
       end)
 
-      opts = Keyword.put(base_opts, :retry, max_retries: 3)
+      {:ok, opts} = GoogleApiClient.init(opts)
+      assert GoogleApiClient.acknowledge(["1", "2"], opts) == :ok
+
+      assert_received :pong
+      assert_received :pong
+      assert_received :pong
+      refute_received _
+    end
+
+    test "retries with ok response", %{opts: opts} do
+      test_pid = self()
+      max_retries = opts[:retry][:max_retries]
+      {:ok, counter} = Agent.start_link(fn -> 0 end)
+
+      Tesla.Mock.mock(fn %{method: :post} ->
```

Comment:
I'd like to avoid using mocks in the test suite in favour of Bypass since we can find the bugs that triggered this PR in the first place. We don't have to change all the test suite at once, but I'd like to write new tests using it and rewrite existing test when we need to touch them.

WDYT @mcrumm @josevalim?

Reply:
Agreed! :+1:

Reply:
💯